Feature/cv 581 raster habitats

davemfish opened the pull request:
This PR adds support for accepting raster format habitat layers in the CV model. Habitat maps are only used to represent presence/absence of habitat, so for rasters, I chose to treat 0 and nodata as absence, and anything else as presence.

The new `search_for_raster_habitat` function makes use of the pre-existing utility `_aggregate_raster_values_in_radius`. That function just needed to be made slightly more abstract by taking an `aggregation_op` (inspired by raster calculator) so that it can be shared by the `relief`, `population`, and `habitat` calculations.

I updated test habitat data to replace a shapefile with a raster. This is hard to notice in the test code because that reference is hidden inside a CSV. But `test_habitat_rank` and `test_complete_run` now run with a raster habitat layer. And I updated the sample data in the same manner.

Fixes #581 and fixes #628 

# Checklist
- [x] Updated HISTORY.rst (if these changes are user-facing)

- [x] Updated the user's guide (if needed) -- only a minor update was needed because the details in the Data Needs section now come straight from the `ARGS_SPEC`!

phargogh commented on src/natcap/invest/coastal_vulnerability.py:
```diff
@@ -1885,21 +1930,103 @@ def _schedule_habitat_tasks(
-                  target_habitat_pickle_path),
-            target_path_list=[target_habitat_pickle_path],
-            task_name='searching for %s' % habitat_row.id))
+        vector = gdal.OpenEx(base_habitat_path, gdal.OF_VECTOR)
+        if vector:
+            vector = None
+            habitat_task_list.append(task_graph.add_task(
+                func=search_for_vector_habitat,
+                args=(base_shore_point_vector_path,
+                      habitat_row.distance,
+                      habitat_row.rank,
+                      habitat_row.id,
+                      base_habitat_path,
+                      target_habitat_pickle_path),
+                target_path_list=[target_habitat_pickle_path],
+                task_name='searching for %s' % habitat_row.id))
+            continue
+        raster = gdal.OpenEx(base_habitat_path, gdal.OF_RASTER)
+        if raster:
+            raster = None
```
For the spatial type checking (raster/vector), this sounds like a good use case for `pygeoprocessing.get_gis_type()`.  It doesn't do much more than what you have here with the conditionals, but would mean that you wouldn't need to remember to `None` out the raster/vector variables.

davemfish replied:
Cool, I did not know that existed

phargogh commented on src/natcap/invest/coastal_vulnerability.py:
```diff
@@ -2985,10 +3078,14 @@ def _validate_habitat_table_paths(habitat_table_path):
         try:
             pygeoprocessing.get_vector_info(base_habitat_path)
         except ValueError:
-            bad_paths.append(base_habitat_path)
+            try:
+                pygeoprocessing.get_raster_info(base_habitat_path)
+            except ValueError:
+                bad_paths.append(base_habitat_path)
```
This might be another use case for `pygeoprocessing.get_gis_type`, since `get_raster_info` and `get_vector_info` will both return a bunch of other information that aren't strictly needed.

phargogh commented on src/natcap/invest/coastal_vulnerability.py:
```diff
@@ -2599,27 +2724,24 @@ def _aggregate_raster_values_in_radius(
-            for sample_distance and raster geotransform.
         target_pickle_path (string): path to pickle file storing dict
             keyed by point id.
+        aggregation_op (function): takes a signal array and a mask array
+            of same shape. Returns scalar.
```
Or returns `numpy.nan` if no masked signal pixels are found?

davemfish replied:
Not strictly. The habitat op actually returns 5 in that case, representing no habitat found. Because we're allowing nodata to represent known absence of habitat. As I'm writing that out, it starts to sound like a bad idea. Hmm.

davemfish replied:
Sorry I was splitting hairs here. I added "or numpy.nan". But I think we can leave it up to the actual `op` to describe when it might return `numpy.nan`, as it might vary by the `op`.

phargogh commented on src/natcap/invest/coastal_vulnerability.py:
```diff
@@ -2669,78 +2790,50 @@ def _aggregate_raster_values_in_radius(
+        if nodata is not None:
+            kernel_mask &= ~numpy.isclose(array, nodata)
+
+        result[shore_id] = aggregation_op(array, kernel_mask)
```
I see that in the removed code, there's a check in the `'density'` op about whether there are `nan` pixels.  Is it expected to have `result[shore_id] = numpy.nan`, for the case that one of the ops does end up returning `nan`?

davemfish replied:
Yes, `nan` values are expected to exist in those `result` dicts, if that is your question.

davemfish replied:
There are tests to make sure `nan` is handled properly later on when these intermediate results are used in subsequent equations.

phargogh reviewed (changes requested):
Thanks, Dave!  I just had a couple small things I was wondering about.

davemfish commented on src/natcap/invest/coastal_vulnerability.py:
```diff
@@ -2669,78 +2790,50 @@ def _aggregate_raster_values_in_radius(
-                sample_distance, os.path.basename(base_raster_path)))
+            # Not sure how we get here, but clamping to 0 allows
+            # the aggregation_op to decide what to return given
+            # an empty array.
```
Hmm I should think on this a bit more. If we're completely off the raster, then maybe even the habitat op should return `nan` instead of the rank representing no habitat protection.

davemfish replied:
I updated this function to do just that.

davemfish commented:
Thanks @phargogh . I pushed some changes to use `pygeoprocessing.get_gis_type`. And your other comments are making me want to rethink things a bit. About if/when to assign `nan` as a habitat rank. Or even if we should be more strict with the raster habitat inputs and not allow nodata to represent "known absence" of habitat. Let me think about that.

davemfish commented:
Okay @phargogh I updated the `_aggregate_raster_values_in_radius` to completely bypass the local `aggregation_op` and immediately assign `result[shore_id] = numpy.nan` when the point + radius are completely outside the bounds of the raster.

I updated `test_aggregate_raster_edge_cases` to demonstrate this.

Prior to this change there was no way for a shore point's habitat value to be `nan`, so I updated and tested `_calc_Rhab` to make sure the `nan` propagates correctly. Then `Rhab` is used in the `geometric_mean` for the final exposure calculation, and that function is already setup to handle `nan` correctly.

phargogh reviewed (approved):
Makes sense to me ... thanks, @davemfish !
